File: internal/core/ports/storage.go
package ports

import (
	"context"
	"encoding/json"
	"time"

	"github.com/tjfontaine/polyglot-llm-gateway/internal/core/domain"
)

// ConversationStore defines the interface for conversation storage
type ConversationStore interface {
	// CreateConversation creates a new conversation
	CreateConversation(ctx context.Context, conv *Conversation) error

	// GetConversation retrieves a conversation by ID
	GetConversation(ctx context.Context, id string) (*Conversation, error)

	// AddMessage adds a message to a conversation
	AddMessage(ctx context.Context, convID string, msg *Message) error

	// ListConversations lists conversations with pagination
	ListConversations(ctx context.Context, opts ListOptions) ([]*Conversation, error)

	// DeleteConversation deletes a conversation
	DeleteConversation(ctx context.Context, id string) error

	// Close closes the storage connection
	Close() error
}

// ResponseStore defines the interface for Responses API storage
type ResponseStore interface {
	ConversationStore

	// SaveResponse saves a response for later retrieval
	SaveResponse(ctx context.Context, resp *ResponseRecord) error

	// GetResponse retrieves a response by ID
	GetResponse(ctx context.Context, id string) (*ResponseRecord, error)

	// UpdateResponseStatus updates the status of a response
	UpdateResponseStatus(ctx context.Context, id, status string) error

	// GetResponsesByPreviousID retrieves responses that continue from a given response
	GetResponsesByPreviousID(ctx context.Context, previousID string) ([]*ResponseRecord, error)

	// ListResponses lists responses with pagination
	ListResponses(ctx context.Context, opts ListOptions) ([]*ResponseRecord, error)
}

// Interaction represents a unified view of either a conversation or a response
type Interaction struct {
	ID           string            `json:"id"`
	Type         string            `json:"type"` // "conversation" or "response"
	TenantID     string            `json:"tenant_id"`
	Status       string            `json:"status,omitempty"` // For responses: in_progress, completed, failed, cancelled
	Model        string            `json:"model,omitempty"`
	Metadata     map[string]string `json:"metadata,omitempty"`
	MessageCount int               `json:"message_count,omitempty"` // For conversations
	CreatedAt    time.Time         `json:"created_at"`
	UpdatedAt    time.Time         `json:"updated_at"`
}

// Conversation represents a conversation thread
type Conversation struct {
	ID        string            `json:"id"`
	TenantID  string            `json:"tenant_id"`
	Metadata  map[string]string `json:"metadata,omitempty"`
	Messages  []Message         `json:"messages"`
	CreatedAt time.Time         `json:"created_at"`
	UpdatedAt time.Time         `json:"updated_at"`
}

// Message represents a single message in a conversation
type Message struct {
	ID        string    `json:"id"`
	Role      string    `json:"role"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

// ListOptions defines options for listing conversations
type ListOptions struct {
	TenantID string
	Limit    int
	Offset   int
}

// ResponseRecord represents a stored Responses API response
type ResponseRecord struct {
	ID                 string            `json:"id"`
	TenantID           string            `json:"tenant_id"`
	Status             string            `json:"status"` // "in_progress", "completed", "failed", "cancelled"
	Model              string            `json:"model"`
	Request            json.RawMessage   `json:"request"`  // Serialized request
	Response           json.RawMessage   `json:"response"` // Serialized response
	Metadata           map[string]string `json:"metadata,omitempty"`
	PreviousResponseID string            `json:"previous_response_id,omitempty"`
	CreatedAt          time.Time         `json:"created_at"`
	UpdatedAt          time.Time         `json:"updated_at"`
}

// InteractionStore defines the interface for unified interaction storage.
// This provides full visibility into request/response pairs with canonical mapping data.
type InteractionStore interface {
	ConversationStore

	// SaveInteraction saves an interaction record
	SaveInteraction(ctx context.Context, interaction *domain.Interaction) error

	// AppendInteractionEvent appends an auditable pipeline event for an interaction.
	AppendInteractionEvent(ctx context.Context, event *domain.InteractionEvent) error

	// ListInteractionEvents returns events for an interaction ordered by time.
	ListInteractionEvents(ctx context.Context, interactionID string, opts InteractionListOptions) ([]*domain.InteractionEvent, error)

	// GetInteraction retrieves an interaction by ID
	GetInteraction(ctx context.Context, id string) (*domain.Interaction, error)

	// UpdateInteraction updates an existing interaction
	UpdateInteraction(ctx context.Context, interaction *domain.Interaction) error

	// ListInteractions lists interactions with pagination and optional filtering
	ListInteractions(ctx context.Context, opts InteractionListOptions) ([]*domain.InteractionSummary, error)

	// SetThreadState stores the last response ID for a thread key
	SetThreadState(threadKey, responseID string) error

	// GetThreadState retrieves the last response ID for a thread key
	GetThreadState(threadKey string) (string, error)
}

// InteractionListOptions defines options for listing interactions
type InteractionListOptions struct {
	TenantID      string
	Frontdoor     domain.APIType // Filter by frontdoor type
	Provider      string         // Filter by provider
	Status        string         // Filter by status
	Limit         int
	Offset        int
	InteractionID string // Optional: restrict to a single interaction (for event listing)
}

File: ts/packages/gateway-core/src/ports/storage.ts
/**
 * Storage provider port.
 *
 * @module ports/storage
 */

import type { Message, Usage } from '../domain/types.js';
import type { ShadowResult } from '../domain/shadow.js';
import type { InteractionEvent } from '../domain/events.js';

// ============================================================================
// Conversation Types
// ============================================================================

/**
 * A stored conversation (for Chat Completions API).
 */
export interface Conversation {
    /** Conversation ID. */
    id: string;

    /** Tenant ID. */
    tenantId: string;

    /** App name. */
    appName?: string | undefined;

    /** Model used. */
    model?: string | undefined;

    /** Conversation messages. */
    messages: StoredMessage[];

    /** Metadata. */
    metadata?: Record<string, string> | undefined;

    /** Creation timestamp. */
    createdAt: Date;

    /** Last update timestamp. */
    updatedAt: Date;
}

/**
 * A stored message in a conversation.
 */
export interface StoredMessage {
    /** Message ID. */
    id: string;

    /** Message role. */
    role: 'system' | 'user' | 'assistant' | 'tool';

    /** Message content. */
    content: string;

    /** Token usage (for assistant messages). */
    usage?: Usage | undefined;

    /** Timestamp. */
    timestamp: Date;
}

// ============================================================================
// Response Types (for Responses API)
// ============================================================================

/**
 * A stored response record (for Responses API).
 */
export interface ResponseRecord {
    /** Response ID. */
    id: string;

    /** Tenant ID. */
    tenantId: string;

    /** App name. */
    appName?: string | undefined;

    /** Thread key for linking responses. */
    threadKey?: string | undefined;

    /** Previous response ID. */
    previousResponseId?: string | undefined;

    /** Model used. */
    model: string;

    /** Response status. */
    status: 'in_progress' | 'completed' | 'incomplete' | 'cancelled' | 'failed';

    /** Request data. */
    request?: unknown;

    /** Response data. */
    response?: unknown;

    /** Error if failed. */
    error?: unknown;

    /** Token usage. */
    usage?: Usage | undefined;

    /** Metadata. */
    metadata?: Record<string, string> | undefined;

    /** Creation timestamp. */
    createdAt: Date;

    /** Last update timestamp. */
    updatedAt: Date;
}

// ============================================================================
// Interaction Types (Unified)
// ============================================================================

/** Type of interaction. */
export type InteractionType = 'conversation' | 'response';

/**
 * Unified interaction summary for listing.
 */
export interface InteractionSummary {
    /** Interaction ID. */
    id: string;

    /** Interaction type. */
    type: InteractionType;

    /** Tenant ID. */
    tenantId: string;

    /** App name. */
    appName?: string | undefined;

    /** Model used. */
    model?: string | undefined;

    /** Message count (for conversations). */
    messageCount?: number | undefined;

    /** Status (for responses). */
    status?: string | undefined;

    /** Creation timestamp. */
    createdAt: Date;

    /** Last update timestamp. */
    updatedAt: Date;
}

// ============================================================================
// List Options
// ============================================================================

/**
 * Options for listing items.
 */
export interface ListOptions {
    /** Maximum number of items to return. */
    limit?: number | undefined;

    /** Offset for pagination. */
    offset?: number | undefined;

    /** Order by field. */
    orderBy?: string | undefined;

    /** Sort direction. */
    order?: 'asc' | 'desc' | undefined;
}

/**
 * Options for listing interactions.
 */
export interface InteractionListOptions extends ListOptions {
    /** Filter by type. */
    type?: InteractionType | undefined;

    /** Filter by tenant. */
    tenantId?: string | undefined;
}

/**
 * Options for listing divergent shadow results.
 */
export interface DivergenceListOptions extends ListOptions {
    /** Filter by tenant. */
    tenantId?: string | undefined;

    /** Only include structural divergences. */
    structuralOnly?: boolean | undefined;
}

// ============================================================================
// Conversation Store Interface
// ============================================================================

/**
 * Storage for conversations.
 */
export interface ConversationStore {
    /**
     * Saves a conversation.
     */
    saveConversation(conversation: Conversation): Promise<void>;

    /**
     * Gets a conversation by ID.
     */
    getConversation(id: string): Promise<Conversation | null>;

    /**
     * Lists conversations for a tenant.
     */
    listConversations(
        tenantId: string,
        options?: ListOptions,
    ): Promise<Conversation[]>;

    /**
     * Deletes a conversation.
     */
    deleteConversation?(id: string): Promise<void>;
}

// ============================================================================
// Response Store Interface
// ============================================================================

/**
 * Storage for responses.
 */
export interface ResponseStore {
    /**
     * Saves a response record.
     */
    saveResponse(response: ResponseRecord): Promise<void>;

    /**
     * Gets a response by ID.
     */
    getResponse(id: string): Promise<ResponseRecord | null>;

    /**
     * Lists responses for a tenant.
     */
    listResponses(
        tenantId: string,
        options?: ListOptions,
    ): Promise<ResponseRecord[]>;

    /**
     * Updates a response.
     */
    updateResponse?(
        id: string,
        updates: Partial<ResponseRecord>,
    ): Promise<void>;

    /**
     * Deletes a response.
     */
    deleteResponse?(id: string): Promise<void>;
}

// ============================================================================
// Interaction Store Interface
// ============================================================================

/**
 * Unified storage for interactions (conversations + responses).
 */
export interface InteractionStore {
    /**
     * Lists all interactions.
     */
    listInteractions(
        options?: InteractionListOptions,
    ): Promise<InteractionSummary[]>;

    /**
     * Gets total interaction count.
     */
    getInteractionCount(options?: InteractionListOptions): Promise<number>;

    /**
     * Saves an interaction event.
     */
    saveEvent(event: InteractionEvent): Promise<void>;

    /**
     * Gets events for an interaction.
     */
    getEvents(interactionId: string): Promise<InteractionEvent[]>;
}

// ============================================================================
// Shadow Store Interface
// ============================================================================

/**
 * Storage for shadow results.
 */
export interface ShadowStore {
    /**
     * Saves a shadow result.
     */
    saveShadowResult(result: ShadowResult): Promise<void>;

    /**
     * Gets shadow results for an interaction.
     */
    getShadowResults(interactionId: string): Promise<ShadowResult[]>;

    /**
     * Gets a shadow result by ID.
     */
    getShadowResult(id: string): Promise<ShadowResult | null>;

    /**
     * Lists divergent shadow results.
     */
    listDivergentShadowResults(
        options?: DivergenceListOptions,
    ): Promise<ShadowResult[]>;
}

// ============================================================================
// Thread State Store Interface
// ============================================================================

/**
 * Storage for thread state (Responses API continuation).
 */
export interface ThreadStateStore {
    /**
     * Sets the thread state (links thread key to response ID).
     */
    setThreadState(threadKey: string, responseId: string): Promise<void>;

    /**
     * Gets the thread state (latest response ID for thread).
     */
    getThreadState(threadKey: string): Promise<string | null>;
}

// ============================================================================
// Thread Store Interface (OpenAI Assistants-style)
// ============================================================================

/**
 * A stored thread.
 */
export interface StoredThread {
    /** Thread ID. */
    id: string;

    /** Tenant ID. */
    tenantId: string;

    /** Messages in the thread. */
    messages: StoredMessage[];

    /** Metadata. */
    metadata?: Record<string, string> | undefined;

    /** Creation timestamp. */
    createdAt: Date;

    /** Last update timestamp. */
    updatedAt: Date;
}

/**
 * Storage for threads (OpenAI Assistants-style API).
 */
export interface ThreadStore {
    /**
     * Creates a new thread.
     */
    createThread(thread: StoredThread): Promise<void>;

    /**
     * Gets a thread by ID.
     */
    getThread(id: string): Promise<StoredThread | null>;

    /**
     * Adds a message to a thread.
     */
    addMessage(threadId: string, message: StoredMessage): Promise<void>;

    /**
     * Lists messages in a thread.
     */
    listMessages(threadId: string, options?: ListOptions): Promise<StoredMessage[]>;

    /**
     * Deletes a thread.
     */
    deleteThread?(id: string): Promise<void>;
}

// ============================================================================
// Combined Storage Provider Interface
// ============================================================================

/**
 * Full storage provider combining all storage interfaces.
 */
export interface StorageProvider
    extends ConversationStore,
    ResponseStore,
    InteractionStore,
    ShadowStore,
    ThreadStateStore,
    Partial<ThreadStore> {
    /**
     * Closes the storage connection.
     */
    close?(): Promise<void>;
}
